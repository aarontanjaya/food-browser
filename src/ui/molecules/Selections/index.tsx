import Chip from '@app/ui/atoms/Chip';
import Tab from '@app/ui/atoms/Tab';
import { cn } from '@utils';

type SelectionsProps = {
  items: SelectionItem[];
  value: string | number;
  onChange: (value: string | number) => void;
  className?: string;
  itemClassName?: string;
};

type SelectionItem = {
  label: React.ReactNode;
  value: string | number;
};

export default function Selections({ items, value, onChange, className }: SelectionsProps) {
  return (
    <div className={cn('scrollbar-hidden flex flex-row overflow-x-auto', className)}>
      <div className="hidden flex-row md:flex">
        {items.map((item) => (
          <Tab
            className={cn(
              'border-green-neon-300 hidden rounded-none border border-r border-l-0 first:rounded-l-lg first:border-l last:rounded-r-lg md:block',
            )}
            key={item.value}
            value={item.value}
            isActive={item.value === value}
            onClick={() => onChange(item.value)}
          >
            {item.label}
          </Tab>
        ))}
      </div>
      <div className="flex flex-row gap-x-3 md:hidden">
        {items.map((item) => (
          <Chip
            className={cn('border-green-neon-300 line-clamp-1 shrink-0 text-nowrap md:hidden')}
            key={item.value}
            value={item.value}
            isActive={item.value === value}
            onClick={() => onChange(item.value)}
          >
            {item.label}
          </Chip>
        ))}
      </div>
    </div>
  );
}
