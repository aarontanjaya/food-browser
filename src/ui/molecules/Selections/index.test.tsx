import Selections from '.';
import { fireEvent, render } from '@testing-library/react';
import { act } from 'react';

const mockItems = [
  { label: 'Item 1', key: 'key1', value: 'value1' },
  { label: 'Item 2', key: 'key2', value: 'value2' },
];
describe('Selections', () => {
  const mockOnChange = jest.fn();
  it('should render correctly without error and have required components', () => {
    // #region ARRANGE
    const screen = render(
      <Selections data-testid="selection" items={mockItems} value={mockItems[0].value} onChange={mockOnChange} />,
    );
    // #endregion

    // #region ASSERT
    expect(screen).toMatchSnapshot();
    // #endregion
  });

  it('should call onChange with correct value when tab is clicked', () => {
    // #region ARRANGE
    const screen = render(
      <Selections data-testid="selection" items={mockItems} value={null} onChange={mockOnChange} />,
    );
    const firstTab = screen.getByTestId(`selection:tab:${mockItems[0].key}:container`);
    // #endregion

    // #region ACT
    act(() => {
      fireEvent.click(firstTab);
    });
    // #endregion

    // #region ASSERT
    expect(mockOnChange).toHaveBeenCalledWith(mockItems[0].value);
    // #endregion
  });

  it('should call onChange with correct value when chip is clicked', () => {
    // #region ARRANGE
    const screen = render(
      <Selections data-testid="selection" items={mockItems} value={null} onChange={mockOnChange} />,
    );
    const secondChip = screen.getByTestId(`selection:chip:${mockItems[1].key}:container`);
    // #endregion

    // #region ACT
    act(() => {
      fireEvent.click(secondChip);
    });
    //#endregion

    // #region ASSERT
    expect(mockOnChange).toHaveBeenCalledWith('value2');
    // #endregion
  });
});
