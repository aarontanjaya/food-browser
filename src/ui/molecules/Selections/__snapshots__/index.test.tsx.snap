// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selections should render correctly without error and have required components 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="scrollbar-hidden flex flex-row overflow-x-auto"
        data-testid="selection:container"
      >
        <div
          class="hidden flex-row md:flex"
          data-testid="selection:tab:container"
        >
          <button
            class="data-[active=true]:bg-green-neon-100 data-[active=true]:text-green-neon-900 bg-white px-6 py-3 text-neutral-600 hover:cursor-pointer border-green-neon-300 hidden rounded-none border border-r border-l-0 first:rounded-l-lg first:border-l last:rounded-r-lg md:block"
            data-active="true"
            data-testid="selection:tab:key1:container"
          >
            Item 1
          </button>
          <button
            class="data-[active=true]:bg-green-neon-100 data-[active=true]:text-green-neon-900 bg-white px-6 py-3 text-neutral-600 hover:cursor-pointer border-green-neon-300 hidden rounded-none border border-r border-l-0 first:rounded-l-lg first:border-l last:rounded-r-lg md:block"
            data-active="false"
            data-testid="selection:tab:key2:container"
          >
            Item 2
          </button>
        </div>
        <div
          class="flex flex-row gap-x-3 md:hidden"
          data-testid="selection:chip:container"
        >
          <button
            class="data-[active=true]:bg-green-neon-100 data-[active=true]:text-green-neon-900 rounded-l-full rounded-r-full bg-white px-6 py-3 text-neutral-600 hover:cursor-pointer border-green-neon-300 line-clamp-1 shrink-0 text-nowrap md:hidden"
            data-active="true"
            data-testid="selection:chip:key1:container"
          >
            Item 1
          </button>
          <button
            class="data-[active=true]:bg-green-neon-100 data-[active=true]:text-green-neon-900 rounded-l-full rounded-r-full bg-white px-6 py-3 text-neutral-600 hover:cursor-pointer border-green-neon-300 line-clamp-1 shrink-0 text-nowrap md:hidden"
            data-active="false"
            data-testid="selection:chip:key2:container"
          >
            Item 2
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="scrollbar-hidden flex flex-row overflow-x-auto"
      data-testid="selection:container"
    >
      <div
        class="hidden flex-row md:flex"
        data-testid="selection:tab:container"
      >
        <button
          class="data-[active=true]:bg-green-neon-100 data-[active=true]:text-green-neon-900 bg-white px-6 py-3 text-neutral-600 hover:cursor-pointer border-green-neon-300 hidden rounded-none border border-r border-l-0 first:rounded-l-lg first:border-l last:rounded-r-lg md:block"
          data-active="true"
          data-testid="selection:tab:key1:container"
        >
          Item 1
        </button>
        <button
          class="data-[active=true]:bg-green-neon-100 data-[active=true]:text-green-neon-900 bg-white px-6 py-3 text-neutral-600 hover:cursor-pointer border-green-neon-300 hidden rounded-none border border-r border-l-0 first:rounded-l-lg first:border-l last:rounded-r-lg md:block"
          data-active="false"
          data-testid="selection:tab:key2:container"
        >
          Item 2
        </button>
      </div>
      <div
        class="flex flex-row gap-x-3 md:hidden"
        data-testid="selection:chip:container"
      >
        <button
          class="data-[active=true]:bg-green-neon-100 data-[active=true]:text-green-neon-900 rounded-l-full rounded-r-full bg-white px-6 py-3 text-neutral-600 hover:cursor-pointer border-green-neon-300 line-clamp-1 shrink-0 text-nowrap md:hidden"
          data-active="true"
          data-testid="selection:chip:key1:container"
        >
          Item 1
        </button>
        <button
          class="data-[active=true]:bg-green-neon-100 data-[active=true]:text-green-neon-900 rounded-l-full rounded-r-full bg-white px-6 py-3 text-neutral-600 hover:cursor-pointer border-green-neon-300 line-clamp-1 shrink-0 text-nowrap md:hidden"
          data-active="false"
          data-testid="selection:chip:key2:container"
        >
          Item 2
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
